src/sched/lthread.c

```diff
@@ -857,36 +860,44 @@ int lthread_join(struct lthread* lt, void** ptr, uint64_t timeout)
     struct lthread* current = lthread_get_sched()->current_lthread;
     if (lt->attr.state & BIT(LT_ST_DETACH))
     {
+        SGXLKL_TRACE_THREAD(
+            "[%4d] join (detached): tid=%d count=%d\n",
+            (lthread_self() ? lthread_self()->tid : 0),
+            lt->tid,
+            thread_count);
         return EINVAL;
     }
     _lthread_lock(lt);
     if (lt->attr.state & BIT(LT_ST_EXITED))
     {
         SGXLKL_TRACE_THREAD(
-            "[tid=%-3d] join:  tid=%d count=%d\n",
+            "[%4d] join (exited): tid=%d count=%d\n",
             (lthread_self() ? lthread_self()->tid : 0),
             lt->tid,
             thread_count);
-
-        /* we can test for exited flag only with lock acquired */
-        _lthread_unlock(lt);
     }
     else
     {
         SGXLKL_TRACE_THREAD(
-            "[tid=%-3d] join:  tid=%d count=%d\n",
+            "[%4d] join (waiting): tid=%d count=%d\n",
             (lthread_self() ? lthread_self()->tid : 0),
             lt->tid,
             thread_count);
 
         /* thread is still running, set current lthread as joiner */
-        if (a_cas_p(&lt->lt_join, 0, current) != 0)
+        void* null_ptr = NULL;
+        if (!atomic_compare_exchange_strong(&lt->lt_join, &null_ptr, current) !=
+            0)
         {
             /* there already is a joiner */
             _lthread_unlock(lt);
             return EINVAL;
         }
         _lthread_yield_cb(current, (void*)_lthread_unlock, lt);
+
+        // Reacquire the lthread lock before we start freeing the lthread. It may
+        // still be exiting concurrently.
+        _lthread_lock(lt);
```

Comment:
Isn't the parent thread going to resume execution here sometime after the child thread has exited and the scheduler enqueues its joined thread?

Reply:
I think that the parent thread can resume immediately after it has been added to the scheduler queue here: https://github.com/lsds/sgx-lkl/blob/a1fb63f406557cb02123682efaa48d78018a3421/src/sched/lthread.c#L521 After that line, there are still a number of accesses to `lt` though (and `lt` could have been deallocated if the parent thread runs first before the child thread makes any further progress).